### PR TITLE
ROOM DB and Inventory Integration Pt 2: Electric Boogaloo

### PR DESCRIPTION
**Pull request by lucas-brandt**
Pat merged the first one without my final commit. 👏 for Pat.

In other news:
I added RxRoom architecture and implemented said architecture on the inventory database. Basically, I rewrote how our databases work. Yay.

### CODE REVIEW

File: app/src/main/java/com/example/pbode/smalltown/MainActivity.kt [hunk 1]
```diff
@@ -40,32 +44,34 @@ class MainActivity : AppCompatActivity() {
...
+    //override fun onResume() {
+    //  super.onResume()
+    //  fetchInventory().flatMapIterable { it }
+    //      .subscribe(this::success, this::fail)
+    //}
+
+
+    //TODO: This class may not be the best place for these DB functions, but they will live here for the time being.
+    private fun insertInventoryObject(inventoryData: InventoryDataRecord): Completable {
+        val save = Single.just(inventoryData)
```
Comment (PatB93) [inline app/src/main/java/com/example/pbode/smalltown/MainActivity.kt]: Why not make this 
```
return inventoryDatabase.inventoryDataDao().insert(inventoryData)
        .compose(observableHelper.applySingleSchedulers())
        .ignoreElement()
```

File: app/src/main/java/com/example/pbode/smalltown/database/ObservableHelper.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import io.reactivex.ObservableTransformer
+import io.reactivex.Scheduler
+import io.reactivex.SingleTransformer
+import javax.inject.Inject
+
+open class ObservableHelper @Inject
+constructor(@IoScheduler private val ioScheduler: Scheduler,
+            @MainThreadScheduler private val mainThreadScheduler: Scheduler) {
+
+    open fun <T> applyObservableSchedulers(): ObservableTransformer<T, T> {
```
Comment (PatB93) [inline app/src/main/java/com/example/pbode/smalltown/database/ObservableHelper.kt]: minor, we could make this 
```
open fun <T> applyObservableSchedulers() = ObservableTransformer { it.subscribeOn(ioScheduler).observeOn(mainThreadScheduler) }
```

File: app/src/main/java/com/example/pbode/smalltown/database/ObservableHelper.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        return ObservableTransformer {
+            it.subscribeOn(ioScheduler).observeOn(mainThreadScheduler)
+        }
+    }
+
+    open fun <T> applySingleSchedulers(): SingleTransformer<T, T> {
+        return SingleTransformer {
+            it.subscribeOn(ioScheduler).observeOn(mainThreadScheduler)
```
Comment (PatB93) [inline app/src/main/java/com/example/pbode/smalltown/database/ObservableHelper.kt]: same as above

File: app/src/main/java/com/example/pbode/smalltown/database/inventory/InventoryDataRecord.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import android.arch.persistence.room.Entity
+import android.arch.persistence.room.PrimaryKey
+
+@Entity(tableName = "inventoryData")
+data class InventoryDataRecord(@PrimaryKey(autoGenerate = true) var id: Long? = null,
+                               @ColumnInfo(name = "item") var item: String = "",
+                               @ColumnInfo(name = "cost") var cost: Int = 0,
+                               @ColumnInfo(name = "age") var age: Int? = null,
+                               @ColumnInfo(name = "quantity") var quantity: Int = 0,
+                               @ColumnInfo(name = "image") var image: Int? = null
```
Comment (PatB93) [inline app/src/main/java/com/example/pbode/smalltown/database/inventory/InventoryDataRecord.kt]: for an inventory item it seems strange to have an image be nullable.

File: app/src/main/java/com/example/pbode/smalltown/MainActivity.kt [hunk 5]
```diff
@@ -40,32 +44,34 @@ class MainActivity : AppCompatActivity() {
...
+        return save.map { insertObject(it) }
```
Comment (kpauly2-ford) [inline app/src/main/java/com/example/pbode/smalltown/MainActivity.kt]: This works, but is syntactically odd. You're making a stream out of your input object then mapping it to insert it. Just not how I would put it. It's more clear if you're making the stream out of the action of mapping it, i.e. the first asynchronous block.
```
return Single.fromCallable { insertObject(inventoryData) }....
```

File: app/src/main/java/com/example/pbode/smalltown/database/inventory/InventoryDataRecord.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.example.pbode.smalltown.database.inventory
+
+import android.arch.persistence.room.ColumnInfo
+import android.arch.persistence.room.Entity
+import android.arch.persistence.room.PrimaryKey
+
+@Entity(tableName = "inventoryData")
```
Comment (kpauly2-ford) [inline app/src/main/java/com/example/pbode/smalltown/database/inventory/InventoryDataRecord.kt]: Dunno what specifying tableName does here, I wouldn't do it.

File: app/src/main/java/com/example/pbode/smalltown/dependencyinjection/SmallTownApplication.kt [hunk 7]
```diff
@@ -32,4 +33,9 @@ open class SmallTownApplication : Application(), HasActivityInjector, HasSupport
         return dispatchingAndroidFragmentInjector
     }
 
+    companion object {
+        lateinit var component: ApplicationComponent
```
Comment (kpauly2-ford) [inline app/src/main/java/com/example/pbode/smalltown/dependencyinjection/SmallTownApplication.kt]: Why need?

File: app/src/main/java/com/example/pbode/smalltown/MainActivity.kt [hunk 8]
```diff
@@ -40,32 +44,34 @@ class MainActivity : AppCompatActivity() {
...
-        dbWorkerThread.postTask(task)
+    //TODO: This is an example of how to fetch data from the inventory db, I should actually implement this at some point
+    //override fun onResume() {
+    //  super.onResume()
+    //  fetchInventory().flatMapIterable { it }
+    //      .subscribe(this::success, this::fail)
+    //}
+
+
+    //TODO: This class may not be the best place for these DB functions, but they will live here for the time being.
```
[the remark this answers is not in the document]
Comment (lucas-brandt) [reply app/src/main/java/com/example/pbode/smalltown/MainActivity.kt]: no u

File: app/src/main/java/com/example/pbode/smalltown/MainActivity.kt [hunk 9]
```diff
@@ -40,32 +44,33 @@ class MainActivity : AppCompatActivity() {
         addFragment(landingMenuFragment, R.id.fragment_content)
         replaceFragmentSubject.subscribe { replaceFragment(it, R.id.fragment_content) }
 
-        dbWorkerThread = DbWorkerThread("dbWorkerThread")
-        dbWorkerThread.start()
+        //TODO: This is an example of how to save data into the inventory db, I should actually implement this at some point
```
Comment (PatB93) [inline app/src/main/java/com/example/pbode/smalltown/MainActivity.kt]: Please create a story in Trello and link it.

File: app/src/main/java/com/example/pbode/smalltown/MainActivity.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (PatB93) [reply app/src/main/java/com/example/pbode/smalltown/MainActivity.kt]: Now that we have Trello we can create chores/change requests 😸

File: app/src/main/java/com/example/pbode/smalltown/database/inventory/InventoryDatabase.kt [hunk 10]
```diff
@@ -4,7 +4,7 @@ import android.arch.persistence.room.Database
 import android.arch.persistence.room.RoomDatabase
 import javax.inject.Singleton
 
-@Database(entities = [InventoryDataTable::class], version = 6)
+@Database(entities = [InventoryDataRecord::class], version = 1)
```
Comment (PatB93) [inline app/src/main/java/com/example/pbode/smalltown/database/inventory/InventoryDatabase.kt]: 👍

**Pull request merged**